cluster-autoscaler/cloudprovider/azure/azure_scale_set.go

```diff
@@ -565,106 +663,107 @@ func (scaleSet *ScaleSet) Nodes() ([]cloudprovider.Instance, error) {
-	vms, rerr := scaleSet.GetFlexibleScaleSetVms()
+func (scaleSet *ScaleSet) buildScaleSetCacheForUniform() error {
+	klog.V(3).Infof("updateInstanceCache: resetting instance Cache for scaleSet %s",
+		scaleSet.Name)
+	splay := rand.New(rand.NewSource(time.Now().UnixNano())).Intn(scaleSet.instancesRefreshJitter + 1)
+	lastRefresh := time.Now().Add(-time.Second * time.Duration(splay))
+	vms, rerr := scaleSet.GetScaleSetVms()
 	if rerr != nil {
 		if isAzureRequestsThrottled(rerr) {
-			// Log a warning and update the instance refresh time so that it would retry after cache expiration
-			klog.Warningf("GetFlexibleScaleSetVms() is throttled with message %v, would return the cached instances", rerr)
+			// Log a warning and update the instance refresh time so that it would retry later.
+			// Ensure to retry no sooner than rerr.RetryAfter
+			klog.Warningf("updateInstanceCache: GetScaleSetVms() is throttled with message %v, would return the cached instances", rerr)
+			nextRefresh := lastRefresh.Add(scaleSet.instancesRefreshPeriod)
+			if nextRefresh.Before(rerr.RetryAfter) {
+				delay := rerr.RetryAfter.Sub(nextRefresh)
+				lastRefresh = lastRefresh.Add(delay)
+			}
 			scaleSet.lastInstanceRefresh = lastRefresh
 			return nil
 		}
 		return rerr.Error()
 	}
 
-	scaleSet.instanceCache = buildInstanceCache(vms)
+	instances := []cloudprovider.Instance{}
+	// Note that the GetScaleSetVms() results is not used directly because for the List endpoint,
+	// their resource ID format is not consistent with Get endpoint
+	for i := range vms {
+		// The resource ID is empty string, which indicates the instance may be in deleting state.
+		if *vms[i].ID == "" {
+			continue
+		}
+		resourceID, err := convertResourceGroupNameToLower(*vms[i].ID)
+		if err != nil {
+			// This shouldn't happen. Log a warning message for tracking.
+			klog.Warningf("updateInstanceCache: buildInstanceCache.convertResourceGroupNameToLower failed with error: %v", err)
+			continue
+		}
+
+		instances = append(instances, cloudprovider.Instance{
+			Id:     azurePrefix + resourceID,
+			Status: scaleSet.instanceStatusFromVM(&vms[i]),
+		})
+	}
+
+	scaleSet.instanceCache = instances
 	scaleSet.lastInstanceRefresh = lastRefresh
 
 	return nil
 }
 
 // Note that the GetScaleSetVms() results is not used directly because for the List endpoint,
 // their resource ID format is not consistent with Get endpoint
-func buildInstanceCache(vmList interface{}) []cloudprovider.Instance {
-	instances := []cloudprovider.Instance{}
-
-	switch vms := vmList.(type) {
-	case []compute.VirtualMachineScaleSetVM:
-		for _, vm := range vms {
-			powerState := vmPowerStateRunning
-			if vm.InstanceView != nil && vm.InstanceView.Statuses != nil {
-				powerState = vmPowerStateFromStatuses(*vm.InstanceView.Statuses)
-			}
-			addInstanceToCache(&instances, vm.ID, vm.ProvisioningState, powerState)
-		}
-	case []compute.VirtualMachine:
-		for _, vm := range vms {
```

Comment:
These 2 conditions are the same, thus removing the switch is justified and no impact